main.go

```diff
@@ -245,21 +262,42 @@ func main() {
 	client := &http.Client{
 		Timeout: time.Second * 60,
 	}
-	for _, request := range cowpokeRequests {
-		response, err := client.Do(request)
-		if err != nil {
-			fmt.Println("error executing request:", response, err)
-			os.Exit(0)
-		}
-		contents, err := ioutil.ReadAll(response.Body)
+	var goroutines [](chan bool)
+	for index, request := range cowpokeRequests {
+		done := make(chan bool, 1)
+		go doRequest(catalogCheckRequests[index], request, client, done)
+		goroutines = append(goroutines, done)
+	}
+	for _, routine := range goroutines {
+		<-routine
+	}
+	fmt.Println("... Finished drone-rancher-catalog")
+}
+
+func doRequest(check *http.Request, upgrade *http.Request, client *http.Client, done chan<- bool) {
+	catalogIsThere := false
+	for !catalogIsThere && check != nil {
+		time.Sleep(50 * time.Millisecond)
+		response, err := client.Do(check)
 		if err != nil {
-			fmt.Println("error reading response:", err)
+			fmt.Printf(err.Error())
+			break
 		}
-		fmt.Println("response status code:", response.StatusCode)
-		fmt.Println("content:", string(contents))
-		response.Body.Close()
+		catalogIsThere = response.StatusCode != 404
 	}
-	fmt.Println("... Finished drone-rancher-catalog")
+	response, err := client.Do(upgrade)
+	if err != nil {
+		fmt.Println("error executing request:", response, err)
+		os.Exit(0)
```

Comment:
I do not quite understand you. The outline is currently to run a goroutine for each upgrade call, and then wait for them all to finish. Each goroutine waits for the catalog to be available from github and then calls cowpoke for that catalog. I do not understand how your approach would be different because it is mapping out the tasks to goroutines and then waiting for their completion?